### ISSUE TITLE
problem with logging in the gitlab container

### ISSUE THREAD

**Issue by jave** [sameersbn/docker-gitlab#488]
I get
gitlab_1     |  WARNING: no logs are available with the 'journald' log driver
and later I get
ci_gitlab_1 exited with code 1

the redis and pg containers seem to log properly.

Im using docker-compose, and it is problematic restarting the stack, and it seems I cant get logs either..

this is the compose file
```yaml
postgresql:
  image: quay.io/sameersbn/postgresql:9.4-5
  environment:
    - DB_USER=gitlab
    - DB_PASS=password
    - DB_NAME=gitlabhq_production
  volumes:
    - ./data/gitlab/postgresql:/var/lib/postgresql
redis:
  image: quay.io/sameersbn/redis:latest
  volumes:
    - ./data/gitlab/redis:/var/lib/redis

gitlab:
  image: quay.io/sameersbn/gitlab:8.1.2
  links:
    - redis:redisio
    - postgresql:postgresql
  ports:
    - "50003:80"
    - "50004:22"
  environment:
    - TZ=Europe/Stockholm
    - GITLAB_TIMEZONE=Stockholm

    - GITLAB_SECRETS_DB_KEY_BASE=long-and-random-alphanumeric-string

    - GITLAB_HOST=localhost
    - GITLAB_PORT=50003
    - GITLAB_SSH_PORT=50004

    - GITLAB_NOTIFY_ON_BROKEN_BUILDS=true
    - GITLAB_NOTIFY_PUSHER=false

    - GITLAB_EMAIL=notifications@example.com
    - GITLAB_EMAIL_REPLY_TO=noreply@example.com
    - GITLAB_INCOMING_EMAIL_ADDRESS=reply@example.com

    - GITLAB_BACKUPS=daily
    - GITLAB_BACKUP_TIME=01:00

    - SMTP_ENABLED=true
    - SMTP_DOMAIN=verona.se
    - SMTP_HOST=127.0.0.1
    - SMTP_PORT=587
#    - SMTP_USER=mailer@example.com
#    - SMTP_PASS=password
#    - SMTP_STARTTLS=true
#    - SMTP_AUTHENTICATION=login

    - IMAP_ENABLED=false
    - IMAP_HOST=imap.gmail.com
    - IMAP_PORT=993
    - IMAP_USER=mailer@example.com
    - IMAP_PASS=password
    - IMAP_SSL=true
    - IMAP_STARTTLS=false
    - DEBUG_ENTRYPOINT=true
  volumes:
    - ./data/gitlab/gitlab:/home/git/data
```

**Comment by sameersbn** [comment]
seems like you are using journald logging driver. see https://docs.docker.com/reference/logging/journald/. Please check.

**Comment by jave** [comment]
I did remove the container and the image for gitlab. After pulling and restarting i get the same problem.

This is with fairly new bits, docker 1.8 and matching docker compose.

I'm suspecting some form of race condition that happens only when restarting the gitlab container stack. Starting from fresh seems to nearly always work.

**Comment by jave** [comment]
I split out the gitlab container from the redis and pg ones, which
still run in the compose file.

start the gitlab container like this, now with the json-file logger.

docker run --name gitlab --log-driver=json-file \
    --link ci_postgresql_1:postgresql --link ci_redis_1:redisio \
    --publish 50004:22 --publish 50003:80 \
    --env 'GITLAB_PORT=50003' --env 'GITLAB_SSH_PORT=50004' \
    --env 'GITLAB_SECRETS_DB_KEY_BASE=long-and-random-alpha-numeric-string' \
    --volume ./data/gitlab/gitlab:/home/git/data \
    quay.io/sameersbn/gitlab:8.1.2

> > > > > > Waiting for database server to accept connections...........................................................[root@localhost ci]# [root@localhost ci]# 

so gitlab cant connect to postgres it seems.

but i can enter the pg container and issue psql queries.

su postgres
psql
\c gitlabhq_production
select * from projects;

lists the test projects so postgres seems happy to a degree anyway.

so, theres some other reason gitlab cant connect to the pg container.
I wasnt able to verify network connectivity to the pg 5432 port yet,
because im not sure what the params should look like.

this didnt work.

psql -h 127.0.0.1 -p 5432 gitlabhq_production
Password: 
psql: FATAL:  password authentication failed for user "postgres"

**Comment by jave** [comment]
this works from the host(entering password at prompt):
psql -U gitlab -h 172.17.0.23 gitlabhq_production

**Comment by jave** [comment]
this is from inside the gitlab container

psql -U gitlab -h 172.17.0.23 gitlabhq_production
psql: could not connect to server: No route to host
	Is the server running on host "172.17.0.23" and accepting
	TCP/IP connections on port 5432?

but ping works.

ping 172.17.0.23
ping 172.17.0.23
PING 172.17.0.23 (172.17.0.23) 56(84) bytes of data.
64 bytes from 172.17.0.23: icmp_seq=1 ttl=64 time=0.110 ms
64 bytes from 172.17.0.23: icmp_seq=2 ttl=64 time=0.093 ms

**Comment by sameersbn** [comment]
@jave it could be that something on the host is not allowing the connection. maybe `-icc=true` is set on the docker daemon due to which the two containers are not allowed to communicated. see https://docs.docker.com/articles/networking/.

If that is not the case, then maybe selinux or some other thing is not allowing the communication .

**Comment by jave** [comment]
On Fedora, the docker daemon gets these flags

/usr/bin/docker daemon --selinux-enabled --log-driver=journald

selinux is in permissive state

The other containers seem to be able to talk to each other.

cant there be some problem in the postgres conf, such that it only allows certain ip ranges?

**Comment by jave** [comment]
the last part of the container pghba conf looks like below.
it seems like the listening row is added for each restart maybe?

# TYPE  DATABASE        USER            ADDRESS                 METHOD

# "local" is for Unix domain socket connections only
local   all             all                                     peer
# IPv4 local connections:
host    all             all             127.0.0.1/32            md5
# IPv6 local connections:
host    all             all             ::1/128                 md5
# Allow replication connections from localhost, by a user with the
# replication privilege.
#local   replication     postgres                                peer
#host    replication     postgres        127.0.0.1/32            md5
#host    replication     postgres        ::1/128                 md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5
host    all             all             0.0.0.0/0               md5

**Comment by jave** [comment]
adding
  log_driver: "json-file"
to each container i get much better logging!

before the gitlab container dies it repeats this statement approx 50 times
gitlab_1     | + /usr/lib/postgresql/9.4/bin/pg_isready -h 172.17.0.37 -p 5432 -U gitlab -d gitlabhq_production -t 1

the same statement from the host works
pg_isready -h 172.17.0.37 -p 5432 -U gitlab -d gitlabhq_production -t 1
172.17.0.37:5432 - accepting connections

**Comment by sameersbn** [comment]
> gitlab_1 | + /usr/lib/postgresql/9.4/bin/pg_isready -h 172.17.0.37 -p 5432 -U gitlab -d gitlabhq_production -t 1

This is a loop that check the connection to the database server. It tries 60 times with an interval of 1 sec between each attempt. This is ok..

The problem should most probably be related to selinux. I can be such a bitch sometimes :anger:.

@pnelsonsr can you help please? I am sure you have come across this issue on fedora.
@jave can you please state your fedora version, docker version, etc.

**Comment by jave** [comment]
[root@localhost ci]# getenforce
Permissive
[root@localhost ci]# docker --version
Docker version 1.8.2-fc23, build 1b3b127-dirty
[root@localhost ci]# docker-compose version
docker-compose version: 1.5.0rc2
docker-py version: 1.5.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
[root@localhost ci]# uname -a
Linux localhost.localdomain 4.2.0-300.fc23.x86_64 #1 SMP Fri Sep 4 13:27:08 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
[root@localhost ci]# cat /etc/fedora-release 
Fedora release 23 (Twenty Three)

**Comment by jave** [comment]
I have a solution now, which sucks, but it works.
- I had to export both the redis port and the postgres port to the host
- I had to use the external postgres/redis conf options, except redis and pg are running in containers rather than being external

- I also had to chown the gitlab repo dir, for some reason. they were owned by root. Im not sure why that happened.

heres the working compose file. 10.0.75.24 is the ip of the host.

postgresql:
  log_driver: "json-file"
  image: quay.io/sameersbn/postgresql:9.4-5
  ports:
    - "50005:5432"  
  environment:
    - DB_USER=gitlab
    - DB_PASS=password
    - DB_NAME=gitlabhq_production
  volumes:
    - ./data/gitlab/postgresql:/var/lib/postgresql
redis:
  log_driver: "json-file"  
  image: quay.io/sameersbn/redis:latest
  volumes:
    - ./data/gitlab/redis:/var/lib/redis
  ports:
    - "50006:6379"  

gitlab:
  log_driver: "json-file"  
  image: quay.io/sameersbn/gitlab:8.1.2
  links:
    - redis:redisio
    - postgresql:postgresql
  ports:
    - "50003:80"
    - "50004:22"
  environment:
    - REDIS_HOST=10.0.75.24
    - REDIS_PORT=50006
    - DB_TYPE=postgres
    - DB_HOST=10.0.75.24
    - DB_PORT=50005
    - TZ=Europe/Stockholm
    - GITLAB_TIMEZONE=Stockholm

    - GITLAB_SECRETS_DB_KEY_BASE=long-and-random-alphanumeric-string

    - GITLAB_HOST=localhost
    - GITLAB_PORT=50003
    - GITLAB_SSH_PORT=50004

    - GITLAB_NOTIFY_ON_BROKEN_BUILDS=true
    - GITLAB_NOTIFY_PUSHER=false

    - GITLAB_EMAIL=notifications@example.com
    - GITLAB_EMAIL_REPLY_TO=noreply@example.com
    - GITLAB_INCOMING_EMAIL_ADDRESS=reply@example.com

    - GITLAB_BACKUPS=daily
    - GITLAB_BACKUP_TIME=01:00

    - SMTP_ENABLED=true
    - SMTP_DOMAIN=nwise.se
    - SMTP_HOST=127.0.0.1
    - SMTP_PORT=587
#    - SMTP_USER=mailer@example.com
#    - SMTP_PASS=password
#    - SMTP_STARTTLS=true
#    - SMTP_AUTHENTICATION=login

    - IMAP_ENABLED=false
    - IMAP_HOST=imap.gmail.com
    - IMAP_PORT=993
    - IMAP_USER=mailer@example.com
    - IMAP_PASS=password
    - IMAP_SSL=true
    - IMAP_STARTTLS=false
    - DEBUG_ENTRYPOINT=true
  volumes:
    - ./data/gitlab/gitlab:/home/git/data

**Comment by sameersbn** [comment]
> I had to use the external postgres/redis conf options, except redis and pg are running in containers rather than being external

What do you mean?

> I also had to chown the gitlab repo dir, for some reason. they were owned by root. Im not sure why that happened.

This is selinux doing its thing. Before starting the container did you do,

```bash
mkdir -p <host-volume-path>
sudo chcon -Rt svirt_sandbox_file_t <host-volume-path>
```

This applies some selinux security policies on the host volume. See the readme.

**Comment by pnelsonsr** [comment]
@jave going with the nose bleeding huh (F23)? I have had problems with F22 and docker.  Things like containers not able to resolve ns.  So I always watch the startup of gitlab and if it hangs or errors it usually been problems with name resolution.  It was easy previously when you could virtually mount the systems host file, but after taking away this option (cant remember what release), all my ns troubles started.  So initially I went with a linked bind container.  Which worked for awhile and then it didn't and still does not.  So I have to specifically set the dns of the gitlab container to the IP of my bind container (not the hosts IP).  The bind container is on the same host and is named bind so, I do this with the `--dns=$(docker  inspect --format {{.NetworkSettings.IPAddress}} bind)` parameter.  I tried doing it with a --host=<pg-server>:<ip> and --host=<redis-server>:ip parameters but our environment was more dynamic and we needed it to look at bind for resolution.  So, I'm sorry I don't know specifically about what your seeing but I triggered off your comment about not being able to see the PG container.  And although we do not run PG and Redis as containers, it sounds like container name resolution problems.

Another weird thing is, that we have mutiple backup hosts running the same F22 with matching versions of software installed but I do not have to have the --dns parameter on those.  I've been perplexed by this container resolution problem to say the least.  But with the dns parameter work around, gitlab is working fine.  Here is my environment.

F22 Server
selinux -> enforcing
docker -> 1.8.2-fc22
docker-compose -> not installed
docker-py -> not installed
CPython -> not installed
kernel ->  4.2.3-200.fc22.x86_64
gitlab -> 8.1.2
postgresql -> 9.4.4
redis -> 2.8.22

Again we do not run PG and Redis in containers and do not use docker-compose.  Hope any of this helps.

**Comment by jave** [comment]
@pnelsonsr yes currently I'm trying bleeding nose versions, but I have also tried Centos 7 and Fedora 22. Thanks for the hints!

In my case I use docker compose, and the /etc/hosts manipulation does seem to work, since I can ping the redis and pg container, from within the gitlab container. The odd thing is that the gitlab container cant access the redis and pg ports.

It is possible for the gitlab container to connect to host ports, so if I export pg and redis to the host, and from gitlab accesses these exported container ports on the host, gitlab works.

**Comment by pnelsonsr** [comment]
@jave hmmm on our system that doesn't work (without the --dns= parameter) has PG and Redis on it as host services. Not sure if it matters...   Have you tried to connect directly across the docker network bridge instead of going out through the host 5000x ports?

**Comment by jave** [comment]
@pnelsonsr I tried connecting directly from within the gitlab container to the pg container with the psql tool. That doesnt work. I can ping though. So it feels more like the pg and redis ports doesnt answer on the docker net.

**Comment by sameersbn** [comment]
The problem is that the gitlab container cannot communicate with the redis and postgresql containers. This issue will surely resolved if these services are running directly on the host. Issue is can we figure out why this communication is not happening.

There are two things that come to my mind, selinux or iptables. Can you temporarily disable selinux and check if the communication is allowed. If disabling selinux does not resolve the issue, then it definitely is an iptables thing.

Dockers iptables configuration generally works out of the box, but if you add you own custom rules and stuff it may not work as-is and may require a fair bit of debugging.

Can you try adding the following iptables rules, not sure if it'll work but just give it a shot.

```bash
iptables -A POSTROUTING -s 172.17.0.0/24 ! -d 172.17.0.0/24 -j MASQUERADE
iptables -A POSTROUTING -s 172.17.42.0/24 ! -d 172.17.42.0/24 -j MASQUERADE
```

I would love to try and get this resolved locally, but i dont have access to a F23 system or VM.

**Comment by sameersbn** [comment]
okay you need to follow this issue https://github.com/docker/docker/issues/14396. Should help you get this resolved. If you get it resolved then please mention the resolution here for the benefit of other users.

**Comment by jave** [comment]
@sameersbn thanks for the info! The issue you linked to does indeed seem to be the same I have. So I will test to remove firewalld from the machine when I have physicall access to it, which will take a while unfortunately. I will report here when that happens.

**Comment by pnelsonsr** [comment]
@sameersbn thanks for the issue link.  I read through and was amazed as the docker releases matched my notes.  I looked at my system that had problems with inter-container connectivity and the backup systems that didn't and I found that the system that had problems had firewalld-filesystem and python-firewall installed but firewalld was not running.  So I removed firewalld-filesystem and python-firewall and then did a `dnf update` which had a few updates (including docker-1.8.2-7, docker-selinux-1.8.2-7). After rebooting I started up bind and then I just tried to do my gitlab startup without the `--dns=...` parameter and it worked.  Because I did a few things (and to be honest I have not been trying the gitlab startup without the dns parameter since docker 1.6) I can not say what fixed my problem with inter-container connectivity.  I just happy it works!  Without this thread I would not be "working".  Thanks
@jave our systems have firewalld removed but we had a some remniants of it still installed.  Not sure that it mattered, or that the `dnf update` to the latest docker did it, but my inter-container communication is working again.  I may consider doing PG and Redis in containers if you tell me it works.  And then I may consider your docker-compose.yml as a starting point.  Thanks
